frame/staking/src/lib.rs

```diff
@@ -846,8 +1005,77 @@ decl_module! {
 
 		fn deposit_event() = default;
 
-		fn on_initialize() {
+		/// Does the following:
+		///
+		/// 1. potential storage migration
+		/// 2. sets `ElectionStatus` to `Triggered(now)` where `now` is the block number at which
+		///    the election window has opened. The offchain worker, if applicable, will execute at
+		///    the end of the current block. `submit_election_solution` will accept solutions from
+		///    this block until the end of the era.
+		fn on_initialize(now: T::BlockNumber) {
 			Self::ensure_storage_upgraded();
+			if
+				// if we don't have any ongoing offchain compute.
+				Self::era_election_status() == ElectionStatus::Close &&
+				// and an era is about to be changed.
+				Self::is_current_session_final()
+			{
+				let next_session_change =
+					T::NextSessionChange::estimate_next_session_change(now);
+				if let Some(remaining) = next_session_change.checked_sub(&now) {
+					if remaining <= T::ElectionLookahead::get() && !remaining.is_zero() {
+						// create snapshot.
+						if Self::create_stakers_snapshot() {
+							// Set the flag to make sure we don't waste any compute here in the same era
+							// after we have triggered the offline compute.
+							<EraElectionStatus<T>>::put(
+								ElectionStatus::<T::BlockNumber>::Open(now)
+							);
+							debug::native::info!(
+								target: "staking",
+								"Election window is Open({:?}). Snapshot created",
+								now,
+							);
+						} else {
+							debug::native::warn!(
+								target: "staking",
+								"Failed to create snapshot at {:?}. Election window will remain closed.",
+								now,
+							);
+						}
+
+					}
+				}
+			}
+		}
+
+		/// Check if the current block number is the one at which the election window has been set
+		/// to open. If so, it runs the offchain worker code.
+		fn offchain_worker(now: T::BlockNumber) {
+			debug::RuntimeLogger::init();
+			use offchain_election::{set_check_offchain_execution_status, compute_offchain_election};
+
+			let window_open =
+				Self::era_election_status() == ElectionStatus::<T::BlockNumber>::Open(now);
+
+			if window_open {
+				let offchain_status = set_check_offchain_execution_status::<T>(now);
```

Comment:
@tomusdrw does this have your sign-off as okay? I tested it also locally and it behaved sanely, but ofc I cannot re-produce that far on a testnet.

Reply:
Looks sane!